src/pointer.rs

```diff
@@ -49,11 +49,14 @@ impl core::fmt::Display for Pointer {
     }
 }
 impl Pointer {
-    /// Private constructor for strings that are known to be correctly encoded
+    /// Create a `Pointer` from a string that is known to be correctly encoded.
     ///
-    /// This is a zero-copy constructor
-    fn new<S: AsRef<str> + ?Sized>(s: &S) -> &Self {
-        unsafe { &*(core::ptr::from_ref::<str>(s.as_ref()) as *const Self) }
+    /// This is a cost-free conversion.
+    ///
+    /// ## Safety
+    /// The provided string must adhere to RFC 6901.
```

Comment:
```suggestion
    /// The provided string must adhere to [RFC 6901](https://datatracker.ietf.org/doc/html/rfc6901):
    /// - The pointer must start with `'/'` (%x2F) unless empty
    /// - Tokens must be properly encoded:
    ///     - `'~'` (%x7E) must be escaped as `"~0"`
    ///     - `'/'` (%x2F) must be escaped as `"~1"`
    /// For potentially fallible parsing, see [`Pointer::parse`].
```
I think being explicit here about what's expected may help some folks - or at least persuade them to use something else :). Realistically though, `unsafe` will likely make folks take notice.

Reply:
Yeah, this seems sensible to me, thanks for writing it down.